pkg/kubeconfig/create_kubecfg.go

```diff
@@ -35,42 +35,20 @@ const DefaultKubecfgAdminLifetime = 18 * time.Hour
 func BuildKubecfg(cluster *kops.Cluster, keyStore fi.Keystore, secretStore fi.SecretStore, cloud fi.Cloud, admin time.Duration, configUser string, internal bool, kopsStateStore string, useKopsAuthenticationPlugin bool) (*KubeconfigBuilder, error) {
 	clusterName := cluster.ObjectMeta.Name
 
-	var master string
+	var server string
 	if internal {
-		master = cluster.APIInternalName()
+		server = "https://" + cluster.APIInternalName()
 	} else {
-		master = cluster.Spec.API.PublicName
-		if master == "" {
-			master = "api." + clusterName
+		if cluster.Spec.API.PublicName != "" {
+			server = "https://" + cluster.Spec.API.PublicName
+		} else {
+			server = "https://api." + clusterName
 		}
 	}
 
-	server := "https://" + master
-
-	// We use the LoadBalancer where we know the master DNS name is otherwise unreachable
-	useELBName := false
-
-	// If the master DNS is a gossip DNS name; there's no way that name can resolve outside the cluster
-	if cluster.IsGossip() {
-		useELBName = true
-	}
-
-	// If the cluster has DNS disabled, must use the load balancer name
-	if cluster.UsesNoneDNS() {
-		useELBName = true
-	}
-
-	// If the DNS is set up as a private HostedZone, but here we have to be
-	// careful that we aren't accessing the API over DirectConnect (or a VPN).
-	// We differentiate using the heuristic that if we have an internal ELB
-	// we are likely connected directly to the VPC.
-	privateDNS := cluster.Spec.Networking.Topology != nil && cluster.Spec.Networking.Topology.DNS == kops.DNSTypePrivate
-	internalELB := cluster.Spec.API.LoadBalancer != nil && cluster.Spec.API.LoadBalancer.Type == kops.LoadBalancerTypeInternal
```

Comment:
Are we sure we want to always use internal LBs? Perhaps we should avoid using internal LBs when we have public DNS?

Reply:
How would a cluster with internal LB and public DNS work. Is it possible to configure such a thing?

Reply:
Haven't looked at the code, but I would think access from inside would use LB, whereas access from outside would use DNS.

Reply:
Internal is always DNS, unless you set `API.LoadBalancer.UseForInternalAPI`.
This code is for generating the kubeconfig for the user and for that would expect to use the cert + public DNS name.

Reply:
API validation prevents the combination of NLB + LoadBalancerTypeInternal + UseForInternalAPI

I don't see any cross-validation between DNS topology and load balancer.

Reply:
kOps offers a huge number of config possibilities, with not so much validation.
API LB internal means that the cluster has private topology?
API LB means access to API has to be done via LB. DNS would be a CNAME to the LB?

Reply:
By my reading, if there is a loadbalancer, internal or no, then DNSModelBuilder will point external DNS to the loadbalancer. So for AWS the point is largely moot.